js/adapt-stateful-session.js

```diff
@@ -145,26 +145,35 @@ define([
         'c': courseState,
         'q': componentStates
       };
-      this.printCompletionInformation();
       Adapt.offlineStorage.set(sessionPairs);
+      this.printCompletionInformation(sessionPairs);
     }
 
-    printCompletionInformation() {
-      const courseComplete = Boolean(Adapt.course.get('_isComplete'));
-      const assessmentPassed = Boolean(Adapt.course.get('_isAssessmentPassed'));
+    printCompletionInformation(suspendData) {
+      if (typeof suspendData === 'string') {
+        // In-case LMS data is passed as a string
+        suspendData = JSON.parse(suspendData);
+      }
+      const courseState = SCORMSuspendData.deserialize(suspendData.c);
+      const courseComplete = courseState[0];
+      const assessmentPassed = courseState[1];
       const trackingIdModels = Adapt.data.filter(model => model.get('_type') === this._trackingIdType && model.has('_trackingId'));
       const trackingIds = trackingIdModels.map(model => model.get('_trackingId'));
       if (!trackingIds.length) {
         Adapt.log.info(`course._isComplete: ${courseComplete}, course._isAssessmentPassed: ${assessmentPassed}, ${this._trackingIdType} completion: no tracking ids found`);
         return;
       }
-      const max = Math.max(...trackingIds);
-      const completion = trackingIdModels.reduce((completion, model) => {
-        const trackingId = model.get('_trackingId');
-        completion[trackingId] = model.get('_isComplete') ? '1' : '0';
-        return completion;
-      }, (new Array(max + 1)).join('-').split('')).join('');
-      Adapt.log.info(`course._isComplete: ${courseComplete}, course._isAssessmentPassed: ${assessmentPassed}, ${this._trackingIdType} completion: ${completion}`);
+      const data = SCORMSuspendData.deserialize(suspendData.q);
+      const max = Math.max(...data.map(item => item[0][0]));
+      const completionString = data.reduce((markers, item) => {
+        const trackingId = item[0][0];
+        const mark = item[2][1][0] ? '1' : '0';
```

Comment:
can item[2][1][0] ever result in an undefined error?

Reply:
Not really, if the data is there it will be this line https://github.com/adaptlearning/adapt-contrib-spoor/blob/bb1e3f5fde1bed3da8e7c6291cfcdede7afb6f91/js/serializers/ComponentSerializer.js#L49 or this line https://github.com/adaptlearning/adapt-contrib-spoor/blob/bb1e3f5fde1bed3da8e7c6291cfcdede7afb6f91/js/serializers/ComponentSerializer.js#L61, or through here https://github.com/adaptlearning/adapt-contrib-spoor/blob/bb1e3f5fde1bed3da8e7c6291cfcdede7afb6f91/js/serializers/ComponentSerializer.js#L69 into this line through the booleans https://github.com/adaptlearning/adapt_framework/blob/dd94d83d9e0a76ad05971cd06b709195dc71bd23/src/core/js/models/componentModel.js#L107 and into `_isComplete` as the first trackable boolean https://github.com/adaptlearning/adapt_framework/blob/dd94d83d9e0a76ad05971cd06b709195dc71bd23/src/core/js/models/adaptModel.js#L86

So in all cases it'll be an array inside this kind of structure:
```js
const item = [
  [ trackingId, index], // tracking numbers
  [ hasUserAnswer, isUserAnswerArray, hasAttemptStates ], // structure booleans
  [ // model state
    [ score, attemptsLeft ], // model numbers
    [ isComplete, isInteractionComplete, isSubmitted, isCorrect ], // model booleans
    [ userAnswers ] // sub number/boolean arrays representing the user selections
  ]
];
```

Spoor intentionally converts everything to array of booleans, arrays of numbers or arrays of array in order to compress boolean arrays into a list of binary bits (1bit length), and store numbers with the smallest possible binary footprint for each array, and arrays are just a number of items and a type. [ref](https://github.com/adaptlearning/adapt-contrib-spoor/blob/bb1e3f5fde1bed3da8e7c6291cfcdede7afb6f91/js/serializers/SCORMSuspendData.js)

Reply:
wow. thanks for this info! it has helped my understanding. :)

Reply:
I am receiving the undefined error...
```console
adapt-stateful-session.js:170 Uncaught TypeError: Cannot read property '1' of undefined
    at adapt-stateful-session.js:170
    at Array.reduce (<anonymous>)
    at StatefulSession.printCompletionInformation (adapt-stateful-session.js:168)
    at StatefulSession.saveSessionState (adapt-stateful-session.js:149)
    at a (underscore.min.js:6)
    at underscore.min.js:6
```
```console
item
(2) [Array(2), Array(1)]
0: (2) [1, 0]
1: [false]
lastIndex: (...)
lastItem: (...)
length: 2
__proto__: Array(0)
```
```bash
item.toString()
"1,0,false"
```

Reply:
in which case: yes it is possible @eleanor-heath